File: frontend/src/features/agent-providers/schemas.ts
import { z } from "zod";

export const AgentProviderCapabilitySchema = z.object({
  protocol: z.enum(["codex_chatgpt", "gemini_api", "vertex_ai", "antigravity_cli", "interactions_api"]),
  status: z.enum(["ready", "foundation", "planned"]),
  proxyable: z.boolean(),
  streaming: z.boolean(),
  lifecycleNotes: z.string(),
  operatorAction: z.string(),
  availableUntil: z.string().nullable(),
  notes: z.string(),
});

export const AgentProviderSummarySchema = z.object({
  providerId: z.enum(["codex", "gemini", "antigravity"]),
  displayName: z.string(),
  status: z.enum(["ready", "foundation", "planned"]),
  authModes: z.array(z.enum(["chatgpt_oauth", "api_key", "google_cloud_adc", "cli_keyring"])),
  quotaDimensions: z.array(z.string()),
  dashboardSections: z.array(z.string()),
  capabilities: z.array(AgentProviderCapabilitySchema),
});

export const AgentProviderListSchema = z.object({
  providers: z.array(AgentProviderSummarySchema),
});

export const ProviderOverviewTimeframeSchema = z.enum(["1d", "7d", "30d"]);

export const AgentProviderOverviewItemSchema = z.object({
  providerId: z.enum(["codex", "gemini", "antigravity"]),
  displayName: z.string(),
  status: z.enum(["ready", "foundation", "planned"]),
  accountCount: z.number().int().nonnegative(),
  activeAccountCount: z.number().int().nonnegative(),
  quotaWindowCount: z.number().int().nonnegative(),
  requestCount: z.number().int().nonnegative(),
  successCount: z.number().int().nonnegative(),
  errorCount: z.number().int().nonnegative(),
  inputTokens: z.number().int().nonnegative(),
  outputTokens: z.number().int().nonnegative(),
  cachedInputTokens: z.number().int().nonnegative(),
});

export const AgentProviderOverviewTotalsSchema = z.object({
  providerCount: z.number().int().nonnegative(),
  accountCount: z.number().int().nonnegative(),
  activeAccountCount: z.number().int().nonnegative(),
  quotaWindowCount: z.number().int().nonnegative(),
  requestCount: z.number().int().nonnegative(),
  successCount: z.number().int().nonnegative(),
  errorCount: z.number().int().nonnegative(),
  inputTokens: z.number().int().nonnegative(),
  outputTokens: z.number().int().nonnegative(),
  cachedInputTokens: z.number().int().nonnegative(),
});

export const AgentProviderOverviewSchema = z.object({
  timeframe: ProviderOverviewTimeframeSchema,
  providers: z.array(AgentProviderOverviewItemSchema),
  totals: AgentProviderOverviewTotalsSchema,
});

export type AgentProviderCapability = z.infer<typeof AgentProviderCapabilitySchema>;
export type AgentProviderSummary = z.infer<typeof AgentProviderSummarySchema>;
export type AgentProviderList = z.infer<typeof AgentProviderListSchema>;
export type ProviderOverviewTimeframe = z.infer<typeof ProviderOverviewTimeframeSchema>;
export type AgentProviderOverviewItem = z.infer<typeof AgentProviderOverviewItemSchema>;
export type AgentProviderOverview = z.infer<typeof AgentProviderOverviewSchema>;

File: app/modules/agent_providers/schemas.py
from __future__ import annotations

from typing import Literal

from app.modules.shared.schemas import DashboardModel

AgentProviderId = Literal["codex", "gemini", "antigravity"]
AgentProviderStatus = Literal["ready", "foundation", "planned"]
AgentProviderProtocol = Literal["codex_chatgpt", "gemini_api", "vertex_ai", "antigravity_cli", "interactions_api"]
AgentProviderAuthMode = Literal["chatgpt_oauth", "api_key", "google_cloud_adc", "cli_keyring"]


class AgentProviderCapability(DashboardModel):
    protocol: AgentProviderProtocol
    status: AgentProviderStatus
    proxyable: bool
    streaming: bool
    lifecycle_notes: str
    operator_action: str
    available_until: str | None = None
    notes: str


class AgentProviderSummary(DashboardModel):
    provider_id: AgentProviderId
    display_name: str
    status: AgentProviderStatus
    auth_modes: list[AgentProviderAuthMode]
    quota_dimensions: list[str]
    dashboard_sections: list[str]
    capabilities: list[AgentProviderCapability]


class AgentProviderListResponse(DashboardModel):
    providers: list[AgentProviderSummary]


ProviderOverviewTimeframe = Literal["1d", "7d", "30d"]


class AgentProviderOverviewItem(DashboardModel):
    provider_id: AgentProviderId
    display_name: str
    status: AgentProviderStatus
    account_count: int
    active_account_count: int
    quota_window_count: int
    request_count: int
    success_count: int
    error_count: int
    input_tokens: int
    output_tokens: int
    cached_input_tokens: int


class AgentProviderOverviewTotals(DashboardModel):
    provider_count: int
    account_count: int
    active_account_count: int
    quota_window_count: int
    request_count: int
    success_count: int
    error_count: int
    input_tokens: int
    output_tokens: int
    cached_input_tokens: int


class AgentProviderOverviewResponse(DashboardModel):
    timeframe: ProviderOverviewTimeframe
    providers: list[AgentProviderOverviewItem]
    totals: AgentProviderOverviewTotals
